### ISSUE TITLE
Support Microsoft Azure for the object store

### ISSUE THREAD

**Issue by paraseba** [earth-mover/icechunk#266]

**Comment by kylebarron** [comment]
For both this and https://github.com/earth-mover/icechunk/issues/265, it seems like an `ObjectStore`-based API is ideal, correct?

In https://github.com/developmentseed/object-store-rs I've been developing pyo3 integration for the `object_store` crate, that we could reuse in the icechunk Python APIs. (That repo has both a user-facing Python API _and_ a [Rust-facing pyo3 integration](https://github.com/developmentseed/object-store-rs/tree/main/pyo3-object_store) (not yet well documented)).

I also heard vaguely from @jhamman that you originally used the `object_store` crate for all backends, but then hit some performance issues that made you switch to using `s3` directly. Can you elaborate on that? The `object_store` maintainers are very performance sensitive so I'd 1) be surprised that you could get better performance from using the s3 APIs natively and 2) if so the `object_store` maintainers would probably be interested in your use case. (If preferred, we can move this discussion to a separate thread).

**Comment by geospatial-jeff** [comment]
> We had many random very weird errors with object_store when used with high concurrency. But this was the last week before launch, so we didn't have time to explore. The errors looked like malformed HTTP responses, very weird.

@paraseba I believe this could be due to a difference in how credentialization was implemented in the `object_store` version of the code compared to AWS SDK.  For deployments, object store will fall back to the node profile (credentials provider) when it finds no other form of authentication on the machine.  This is a remote network call (PUT request) to an AWS token endpoint that is sent before each HTTP request to the blob store.  At high concurrencies you may run into rate limits, account-level quotas, and other failures that cause malformed tokens.  I've seen this myself running both `boto3` and `gdal` at big scales.

The original `object_store` implementation would fall back to node profile if it hit this `else` block, and no credentials were configured in the environment - https://github.com/earth-mover/icechunk/blob/e3a119cddfcbe7b38173faeb0b4eb403bd69de7b/icechunk/src/storage/object_store.rs#L92-L94

The AWS SDK **does not** implement this fall back behavior unless you explicitly configure it in the code, and this was not configured when switching the code over to the AWS SDK.  Which would explain why this issue went away with the switch.

This is of course a lot of guesswork based on past experiences, it really depends how the code / dask cluster was deployed. 
 Local cluster on an EC2 instance?  Distributed cluster on k8s?  Running from your local machine?  All have different failure cases when it comes to credentialization.

**Comment by kylebarron** [comment]
I'm interested in implementing this (and potentially #265 at the same time) via the Python bindings. 

Unless you object, I'd like to use [`pyo3-object-store`](https://github.com/developmentseed/obstore/tree/main/pyo3-object_store) for this. I haven't documented it yet, but it's designed as a way to export constructors for the `object_store` crate _once_ with a single API, but reuse them across multiple Rust-Python bindings.

- Publish `pyo3-object_store` to crates.io (now that pyo3-async-runtimes 0.22 has been published)
- Update icechunk to `pyo3` 0.22 and `pyo3-async-runtimes` 0.22
- Export functionality to construct stores from Python.

Any feedback?

**Comment by mpiannucci** [comment]
> Unless you object, I'd like to use [pyo3-object-store](https://github.com/developmentseed/obstore/tree/main/pyo3-object_store) for this. I haven't documented it yet, but it's designed as a way to export constructors for the object_store crate once with a single API, but reuse them across multiple Rust-Python bindings.

Can you explain why it would make sense to use `pyo3-object-store` instead of the `object_store` crate itself in rust? All fo the storage is currently implemented in the rust core, and we will be using this core to support other languages (and platforms) that are not python based and want to make sure we keep that as a core value.

**Comment by kylebarron** [comment]
pyo3-object_store doesn't _replace_ the `object_store` crate. It provides Python builders so that Python users can _create_ `Arc<dyn ObjectStore>` instances.

For example, all of [`storage.rs`](https://github.com/earth-mover/icechunk/blob/5aa9abeae3bb22deffa99ab30aa085233678bc0c/icechunk-python/src/storage.rs) just exports a _very small subset_ of the configuration options to create S3 instances. 

I'll publish and document `pyo3-object_store` tomorrow, but it essentially allows you to

1. [register the builders](https://github.com/developmentseed/obstore/blob/e1ed7659fdea8a0f4bc9eb2e813f6713c125e368/obstore/src/lib.rs#L28). This means that Python users can access the object-store constructors from `icechunk.store`, i.e. `icechunk.store.S3Store` or `icechunk.store.AzureStore`, etc.
2. Then [just accept `PyObjectStore` as a parameter](https://github.com/developmentseed/obstore/blob/e1ed7659fdea8a0f4bc9eb2e813f6713c125e368/obstore/src/delete.rs#L10) in your function exported to Python. [`PyObjectStore::into_inner`](https://github.com/developmentseed/obstore/blob/e1ed7659fdea8a0f4bc9eb2e813f6713c125e368/obstore/src/delete.rs#L12) gives you an `Arc<dyn ObjectStore>`. 

This is **just for the Python binding**. It wouldn't alter any piece of how you already use `object-store` in the main `icechunk` crate.

**Comment by kylebarron** [comment]
> For example, all of [`storage.rs`](https://github.com/earth-mover/icechunk/blob/5aa9abeae3bb22deffa99ab30aa085233678bc0c/icechunk-python/src/storage.rs) just exports a _very small subset_ of the configuration options to create S3 instances.

That is, if you're able to switch back to object-store for S3 as well, you'd be able to delete that entire file by re-exporting the structs from `pyo3-object_store`

**Comment by mpiannucci** [comment]
Got it. So most of that makes sense. Trying to understand how the pyobjectstore configs look in Python land, because our biggest reason for using custom structs is that we all dislike the fsspec method of using keyword args for configuration. So using the structs allows us to strictly type the configuration as best we can. 

I see that the config is a hashmap with specific keys on the rust side, but on the Python side how do those config bindings appear?

**Comment by kylebarron** [comment]
There's strict type hinting for all Python exports. E.g. look at https://github.com/developmentseed/obstore/blob/e1ed7659fdea8a0f4bc9eb2e813f6713c125e368/obstore/python/obstore/store/_aws.pyi. You'd be re-exporting that same class essentially. Or https://github.com/developmentseed/obstore/blob/e1ed7659fdea8a0f4bc9eb2e813f6713c125e368/obstore/python/obstore/store/_azure.pyi for Azure. 

Note that the `config` parameter is a dict with these specified options; there's also [separate options for the HTTP client](https://github.com/developmentseed/obstore/blob/main/obstore/python/obstore/store/_client.pyi) and for the [retry config](https://github.com/developmentseed/obstore/blob/main/obstore/python/obstore/store/_retry.pyi). 

Note that since the underlying `Arc<dyn ObjectStore>` is not ABI-stable, you'd need to export these builders from _your own Python library_. Even though the class names are the same, it wouldn't work to pass in an `obstore.store.AzureStore` to `icechunk-python`. You'd need to instruct users to construct an `icechunk.store.AzureStore`. This is a limitation of dynamic linking with pyo3 for non-ABI-stable objects, see https://github.com/PyO3/pyo3/issues/1444

**Comment by mpiannucci** [comment]
@geospatial-jeff 

> The AWS SDK does not implement this fall back behavior unless you explicitly configure it in the code, and this was not configured when switching the code over to the AWS SDK (https://github.com/earth-mover/icechunk/pull/135). Which would explain why this issue went away with the switch.

It seems the AWS sdk does indeed load the config from env though by default (https://github.com/earth-mover/icechunk/blob/2c9a311c4d69c6cff099eb6474b525193d1ec01c/icechunk/src/storage/s3.rs#L85), so i am confused by this statement. Am i missing something here?

I am responding to this inline here because it is related to the effort to support more stores through object store and s3 being compatible.

**Comment by kylebarron** [comment]
`FromEnv` is probably referring to environment variables, not node instance credentials. Those are two different things.

**Comment by mpiannucci** [comment]
`FromEnv` in the icechunk version does nothing and simply creates an S3 client using the default behavior, which loads from env vars and instance creds etc:

https://github.com/earth-mover/icechunk/blob/2c9a311c4d69c6cff099eb6474b525193d1ec01c/icechunk/src/storage/s3.rs#L76

I tried to find it but ObjectStore seems contrived, if we dont explicitly say we to load from env vars will it work the same way?

**Comment by kylebarron** [comment]
I published an initial beta of `pyo3-object_store` v0.1.0-beta.1: https://crates.io/crates/pyo3-object_store

I'll write more documentation on Monday and then I can lay out my proposal of how the integration would work.

**Comment by mpiannucci** [comment]
Sounds great!

**Comment by paraseba** [comment]
I'd really like it if we could have a single object_store based `Storage` implementation. These are the things I think need to happen for that:
1. Implement and test, currently object_store is used mostly in tests only, so we'll need to gain confidence
2. Implement our virtual chunk management with multiple `ObjectStore` instances. Last time I checked each instance of `ObjectStore` can handle a single bucket (which I hate, but it is what it is)
3. Verify performance. We moved to S3 client because performance was really bad, but [this comment](https://github.com/earth-mover/icechunk/issues/266#issuecomment-2428261096) offers some hope.

To convince ourselves of point 3 above, we can use our dask_write example. This is how we discovered the performance issues before:

```bash
python examples/dask_write.py --url s3://some-bucket/some-prefix
python examples/dask_write.py --url s3://some-bucket/some-prefix  update --t-from 0 --t-to 256 --workers 16  
python examples/dask_write.py --url s3://some-bucket/some-prefix  verify --t-from 0 --t-to 256 --workers 16  
```

As long as object_store can reproduce (close to) the performance of the current code, we are good.

**Comment by kylebarron** [comment]
I personally have quite high confidence in the `object_store` crate because it's used in production by some very big projects, including [the crates.io registry](https://github.com/rust-lang/crates.io/blob/3f48b01afb54705cf97060089595097805ee05cb/Cargo.toml#L92) and any downstream project from [datafusion](https://github.com/apache/datafusion/blob/0458d30b4bf078251ac4c4dfc4669e8beec8bc3b/Cargo.toml#L129). Some really big companies depend on datafusion, like Apple, which recently donated its [datafusion comet](https://github.com/apache/datafusion-comet) project to Apache.

Understandably we need to get more confidence with `object_store` _in icechunk_. 

I see the `artificially_sort_refs_in_mem` flag. Is it crucial for you to get the list result back in sorted order? I don't recall which stores actually do that. I see that ObjectStore [doesn't guarantee the order of results](https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#tymethod.list).

> * Implement our virtual chunk management with multiple `ObjectStore` instances. Last time I checked each instance of `ObjectStore` can handle a single bucket (which I hate, but it is what it is)

This is true, it's connected to a single bucket.

**Comment by paraseba** [comment]
> I see the artificially_sort_refs_in_mem flag. Is it crucial for you to get the list result back in sorted order?

Yes that is crucial. Our in mem store is used for tests and exploration, so we don't expect to have many keys, the artificial sort works there. And with the "normal" object stores, `object_store` preserves the ordering. So I'm not too concerned about this point.

Another similar one is some `object_store` implementation don't handle metadata, which increases the complexity of the code a lot. You'll find many instances of `if self.supports_metadata`.

> Understandably we need to get more confidence with object_store in icechunk.

Yeah .. Icechunk use of the object store could be somewhat particular, we want to read and write many relatively small objects, with very high concurrency and quickly.

**Comment by kylebarron** [comment]
> Another similar one is some `object_store` implementation don't handle metadata

Can you elaborate on this? You're referring to [these attributes](https://docs.rs/object_store/latest/object_store/enum.Attribute.html) in the [`GetResult`](https://docs.rs/object_store/latest/object_store/struct.GetResult.html#structfield.attributes)?

**Comment by kylebarron** [comment]
> You'll find many instances of `if self.supports_metadata`.

I can only find two instances of the string `supports_metadata` on current main
https://github.com/earth-mover/icechunk/blob/fc62650c662b1d1dc9a924571cc45b6652f33758/icechunk/src/storage/object_store.rs#L211-L230

https://github.com/earth-mover/icechunk/blob/fc62650c662b1d1dc9a924571cc45b6652f33758/icechunk/src/storage/object_store.rs#L252-L271

And from the code it looks like [AWS](https://github.com/apache/arrow-rs/blob/e7d370abea4930ce7d58f841885783a279f1ba23/object_store/src/aws/client.rs#L324-L351), [GCS](https://github.com/apache/arrow-rs/blob/e3219705120b52f0d208a043af1caa7d972c7f72/object_store/src/gcp/client.rs#L190-L215), and [Azure](https://github.com/apache/arrow-rs/blob/e3219705120b52f0d208a043af1caa7d972c7f72/object_store/src/azure/client.rs#L196-L217) all support writing attributes. Or maybe it was with local/in-memory stores that you had issues?

**Comment by paraseba** [comment]
Right, the local store is the one not supporting user metadata, you'll see we create it with:
```            
supports_metadata: false,
```

Metadata could become a more important need in the near future, and we don't have an answer for local filesystem support. Thankfully we can still use the in memory store in tests, but it will make things more complicated. For example, to ensure we maintain format compatibility across versions, it would be great to have a local store that supports metadata.

**Comment by paraseba** [comment]
By the way, thank you so much for working on this @kylebarron ! If we manage to make it work it's going to be a big improvement to Icechunk

**Comment by mpiannucci** [comment]
See #411 for WIP PR

**Comment by mpiannucci** [comment]
This is the biggest pain point for using `object_store` in icechunk: https://github.com/apache/arrow-rs/issues/3784 . This is not an actual requirement of the filesystems, just an arbitrary restriction from object_store

**Comment by kylebarron** [comment]
Can you explain why this makes the implementation difficult?

**Comment by mpiannucci** [comment]
We cache filesystems per protocol for virtual references, but we need to change it to cache per bucket instead

**Comment by kylebarron** [comment]
IIRC S3 has separate signing protocols per region, so it's not enough to cache just "s3", you need to always cache a tuple of at least "s3 and region" and probably a few other parameters. At that point I don't caching by bucket is that bad.

**Comment by mpiannucci** [comment]
Good intuition, we have an api that supports that now #502 so well see how well it maps over soon
